feat: Refactored checkbox component

albinAppsmith opened the pull request:
## Description

This pull request introduces a redesigned checkbox component featuring theming support, enhanced keyboard accessibility and an user-friendly API.

Zeplin: https://zpl.io/NwAZw9D

Fixes [#19320](https://github.com/appsmithorg/appsmith/issues/19320)

Depends on #318 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual on storybook 
- Manual on main repo
- Jest
- Cypress

### Test Plan
https://github.com/appsmithorg/TestSmith/issues/2160

### Issues raised during DP testing
https://github.com/appsmithorg/design-system/pull/338#issuecomment-1419104658


## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes


### QA activity:
- [x] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

tanvibhakta commented on packages/design-system-v2/package.json:
```diff
@@ -60,6 +61,7 @@
     "prettier": "^2.8.1",
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
+    "react-stately": "^3.20.0",
```
I thought we made the decision to only install the sub packages we need. Did you mean to install something else here?

albinAppsmith replied:
Fixed this

tanvibhakta replied:
Hmm I don't see the fix?

albinAppsmith replied:
packages/design-system/src/Checkbox/Checkbox.tsx -> line number 2

tanvibhakta commented on packages/design-system-v2/src/Checkbox/Checkbox.stories.tsx:
```diff
@@ -0,0 +1,113 @@
+    },
+    isIndeterminate: {
+      description:
+        "Indicates whether the checkbox is in an indeterminate state.Indeterminism is presentational only.The indeterminate visual representation remains regardless of user interaction.",
```
space after full stop

albinAppsmith replied:
Updated

tanvibhakta commented on packages/design-system-v2/src/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,34 @@
+  const ref = React.useRef(null);
+  const { inputProps } = useCheckbox(props, state, ref);
+  const { focusProps, isFocusVisible } = useFocusRing();
+  const id = `ads-checkbox-${props.value}`;
```
this will not be unique, which might be a problem, yeah?

albinAppsmith replied:
This should be unique, checkbox with same value is not a real use case.

tanvibhakta replied:
Yes, that is what I'm saying in my above comment. the way you're generating id here, there's no guarantee that it will be unique. Should we worry about this? Or should we consider using something like nanoid?

albinAppsmith replied:
We don't need to worry about this,
Developers should not implement more than one checkbox with same value, so until and unless the values are unique this id will also be unique.

tanvibhakta replied:
Okay, just explicitly call this out in the prop doc string please!

tanvibhakta reviewed (commented):
In general this component has a surface that's very different to the one we outlined. Could you update the props on this sheet https://www.notion.so/appsmith/Components-and-their-props-518724e65bfb416e9416b582cb93fc02 to rreflect what we have here, and then take a look at and update any other components that should be consistent with this?

albinAppsmith commented on packages/design-system-v2/src/Checkbox/Checkbox.stories.tsx:
```diff
@@ -0,0 +1,113 @@
+import { Checkbox } from "./Checkbox";
+
+export default {
+  title: "Design System/Checkbox",
```
Why is this needs to be v2?

tanvibhakta replied:
This no longer needs to be V2 😁

albinAppsmith commented on packages/design-system-v2/src/Checkbox/Checkbox.stories.tsx:
```diff
@@ -0,0 +1,113 @@
+  component: Checkbox,
+  argTypes: {
+    name: {
+      description: "The name of the checkbox.",
```
Updated

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+    position: absolute;
+    left: 0;
+    top: var(--ads-v2-spaces-1);
+    width: var(--ads-v2-spaces-5);
```
Let us not use spacing tokens for height and width - our spacing scale is designed to account for gaps between components, not to contain them.

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+const Variables = css`
+  --checkbox-color-label: var(--ads-v2-color-fg);
+  --checkbox-color-border: var(--ads-v2-color-border);
+  --checkbox-color-background-selected: var(--ads-v2-color-bg-brand-secondary);
```
Let's use a single variable for one element - background, or border - and manipulate it within the styles of the class. This way we are not increasing the surface area of our component, and we are following the same pattern for css variable manipulation across components.

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+    top: 11px;
+    width: 5px;
+    height: var(--ads-v2-spaces-1);
+    border-radius: 4px;
```
use the theme variable

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+    top: 9px;
+    width: 10px;
+    height: var(--ads-v2-spaces-1);
+    border-radius: 4px;
```
theme variable

albinAppsmith replied:
This is the checkmark. We don need to define the border radius of this in variable

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+  isChecked: boolean;
+  isIndeterminate?: boolean;
+  isFocusVisible?: boolean;
+  isDisabled?: boolean;
```
I would really like to do isDisabled to follow  the pattern of using the `is` prefix for all booleans, but we've had this discussion and @rohitagarwal88 recommended using the html standard. Given that in certain cases react-aria forces us to use `isDisabled`, and in the interest of keeping a consistent API for our components, can we please adopt this pattern as well? 

This will entail making changes to https://github.com/appsmithorg/design-system/pull/311 to comply.

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.styles.tsx:
```diff
@@ -0,0 +1,165 @@
+  ${({ isDisabled }) => {
+    if (isDisabled) {
+      return css`
+        opacity: 0.5;
```
lets use the theme variable

tanvibhakta commented on packages/design-system/src/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,34 @@
+      isChecked={state.isSelected}
+      isDisabled={isDisabled}
+      isFocusVisible={isFocusVisible}
+      isIndeterminate={isIndeterminate || false}
```
let's add false as a default prop instead? combined with typescript we then won't run into this.

shadabbuchh commented:
Bugs in this PR:

**Functional:**

1. The control `isIntermediate` is wrongly named as `isIndeterminate`.
2. When hovering over the component in disabled state, we do not see the appropriate cursor icon.
3. When `isDisabled` is set to True and the we refresh the page, the component looks interact-able (we get to see the hover state etc.). Also we get to see the CTA cursor.
https://www.loom.com/share/05b6c874117246119b24115cdc216474
4. The `defaultSelected` does not render as expected. Also clicking on the checkbox does not change the state of it instantly, sometime you have to click twice or more that that to change the selection. (This issue is also evident in the Disabled state)
https://www.loom.com/share/006140d276d64310950913f0ea6d151b

**Design:**

1. Line height for the Label is 19px whereas the expected value is 20 px.
2. Letter spacing is -0.24px, expected is -0.1px.
3. Not able to see the focus state.
4. Not able to inspect component in Disabled state. (the checkbox part in particular).

albinAppsmith commented:
Functional:

1. This is correct actually. Sorry by mistake I said it is wrong. [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/:indeterminate)
2. Fixed
3. Fixed
4. Fixed

shadabbuchh commented:
All the above bugs are fixed except the part where `clicking on the checkbox does not change the state of it instantly, sometime you have to click twice or more that that to change the selection.`. Reopening this issue.

albinAppsmith commented:
@shadabbuchh fixed the click issue.

shadabbuchh commented:
All bugs are fixed. This PR is good to go.
